Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm

```diff
@@ -0,0 +1,377 @@
+
+using LocalizationDictionary = NSDictionary<NSString *, NSDictionary *>;
+using PlaceholderDictionary = NSDictionary<NSString *, NSDictionary<NSString *, NSString *> *>;
+
+using namespace WebKit;
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+static LocalizationDictionary *localizationDictionaryAtURL(NSURL *dictionaryFileURL)
+{
+    NSData *dictionaryData = [NSData dataWithContentsOfURL:dictionaryFileURL];
+    if (!dictionaryData)
+        return nil;
+
+    NSError *jsonParsingError;
+    LocalizationDictionary *localizationDictionary = dynamic_objc_cast<NSDictionary>([NSJSONSerialization JSONObjectWithData:dictionaryData options:0 error:&jsonParsingError]);
+
+    if (!localizationDictionary)
+        RELEASE_LOG_ERROR(Extensions , "Could not parse localization dictionary: %{public}@", privacyPreservingDescription(jsonParsingError));
+
+    return localizationDictionary;
+}
+
+@interface _WKWebExtensionLocalization ()
+{
+    LocalizationDictionary *_localizationDictionary;
+    NSString *_localeString;
+    NSLocale *_locale;
+    NSString *_uniqueIdentifier;
+}
+@end
+
+#endif // ENABLE(WK_WEB_EXTENSIONS)
+
+@implementation _WKWebExtensionLocalization
+
++ (BOOL)supportsSecureCoding
+{
+    return YES;
+}
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+- (instancetype)initWithBundleURL:(NSURL *)bundleURL defaultLocale:(NSString *)defaultLocaleString uniqueIdentifier:(NSString *)uniqueIdentifier
+{
+    if (!defaultLocaleString.length) {
+        RELEASE_LOG(Extensions, "Not loading localization for extension bundle at URL %{private}@. No default locale provided.", bundleURL);
+        return [self initWithRegionalLocalization:nil languageLocalization:nil defaultLocalization:nil withBestLocale:nil uniqueIdentifier:uniqueIdentifier];
+    }
+
+    NSLocale *currentLocale = NSLocale.autoupdatingCurrentLocale;
+    NSString *languageCode = currentLocale.languageCode;
+    NSString *countryCode = currentLocale.countryCode;
+
+    NSString *regionalLocaleString = [NSString stringWithFormat:@"%@_%@", languageCode, countryCode];
+    NSURL *defaultLocaleDictionaryURL = [bundleURL URLByAppendingPathComponent:[NSString stringWithFormat:localizationDictionaryJSONPathFormat, defaultLocaleString]];
+    NSURL *languageDictionaryURL = [bundleURL URLByAppendingPathComponent:[NSString stringWithFormat:localizationDictionaryJSONPathFormat, languageCode]];
+    NSURL *regionalDictionaryURL = [bundleURL URLByAppendingPathComponent:[NSString stringWithFormat:localizationDictionaryJSONPathFormat, regionalLocaleString]];
+
+    NSString *bestLocaleString;
+
+    LocalizationDictionary *defaultLocaleDictionary = localizationDictionaryAtURL(defaultLocaleDictionaryURL);
+    if (!!defaultLocaleDictionary)
+        bestLocaleString = defaultLocaleString;
+    else
+        RELEASE_LOG(Extensions, "Could not find localization for %{public}@ for extension bundle at URL %{private}@", defaultLocaleString, bundleURL);
```

Comment:
RELEASE_LOG_ERROR

Reply:
I think `RELEASE_LOG_DEBUG` are fine here, these are not really errors just attempts to find the right file.

Reply:
In fact I think a single `RELEASE_LOG_DEBUG` after all the if lines would be best, logging what locale was picked not what couldn't be found.